File: slidestation/store.py
"""Persistent state: config, sessions (one per tray/box), scans and slide groups."""
from __future__ import annotations

import hashlib
import json
import os
import re
import threading
import time
import uuid
from datetime import datetime
from pathlib import Path

from . import stats
from .imaging import Params

CONFIG_DIR = Path(os.environ.get("SLIDESTATION_HOME", Path.home() / ".slidestation"))
CONFIG_FILE = CONFIG_DIR / "config.json"

DEFAULT_CONFIG = {
    "library": str(Path.home() / "Pictures" / "Slide Station"),
    "immich_url": "",
    "immich_key": "",
    "keep_originals": True,
    "keep_exports": False,
    "learning_enabled": True,
    "jpeg_quality": 95,
    "stats_target": stats.DEFAULT_TARGET,  # slides to digitise in all, for the projected finish
}

lock = threading.RLock()


def _atomic_write(path: Path, data: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(".tmp")
    tmp.write_text(json.dumps(data, indent=1))
    os.replace(tmp, path)


def load_config() -> dict:
    cfg = dict(DEFAULT_CONFIG)
    if CONFIG_FILE.exists():
        cfg.update(json.loads(CONFIG_FILE.read_text()))
    return cfg


def save_config(cfg: dict) -> None:
    _atomic_write(CONFIG_FILE, cfg)
    try:
        os.chmod(CONFIG_FILE, 0o600)  # holds the Immich API key
    except OSError:
        pass


def library() -> Path:
    p = Path(load_config()["library"]).expanduser()
    (p / "sessions").mkdir(parents=True, exist_ok=True)
    return p


def sha1_file(path: str | Path) -> str:
    h = hashlib.sha1()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(1 << 20), b""):
            h.update(chunk)
    return h.hexdigest()


def slugify(s: str) -> str:
    s = re.sub(r"[^\w\-]+", "-", s.strip().lower()).strip("-")
    return s or "session"


# --------------------------------------------------------------------------- imported index (dedupe)


def _index_file() -> Path:
    return library() / "imported.json"


def imported_index() -> dict:
    f = _index_file()
    return json.loads(f.read_text()) if f.exists() else {}


def add_to_index(entries: dict) -> None:
    with lock:
        idx = imported_index()
        idx.update(entries)
        _atomic_write(_index_file(), idx)


# --------------------------------------------------------------------------- presets


def _presets_file() -> Path:
    return library() / "presets.json"


def load_presets() -> list[dict]:
    """Named colour looks, library-wide: [{"name", "params", "created"}], in the order saved."""
    f = _presets_file()
    try:
        return json.loads(f.read_text()).get("presets", []) if f.exists() else []
    except (ValueError, AttributeError):
        return []


def save_presets(presets: list[dict]) -> None:
    with lock:
        _atomic_write(_presets_file(), {"presets": presets})


# --------------------------------------------------------------------------- sessions


class Session:
    """A batch of slides (typically a tray or box) that becomes one Immich album."""

    def __init__(self, sid: str):
        self.id = sid
        self.dir = library() / "sessions" / sid
        self.file = self.dir / "session.json"
        self.data = json.loads(self.file.read_text())

    # paths
    @property
    def originals(self) -> Path:
        return self.dir / "originals"

    @property
    def cache(self) -> Path:
        return self.dir / "cache"

    @property
    def export_dir(self) -> Path:
        return self.dir / "export"

    def original_path(self, scan_id: str) -> Path:
        return self.originals / self.data["scans"][scan_id]["file"]

    @classmethod
    def create(cls, name: str, album: str | None = None, date: str = "") -> "Session":
        sid = time.strftime("%Y%m%d-%H%M%S-") + uuid.uuid4().hex[:4]
        d = library() / "sessions" / sid
        for sub in ("originals", "cache", "export"):
            (d / sub).mkdir(parents=True, exist_ok=True)
        data = {
            "id": sid,
            "name": name or "Untitled tray",
            "album": album if album is not None else (name or "Untitled tray"),
            "date": date,
            "created": time.time(),
            "defaults": Params().to_dict(),
            "scans": {},
            "groups": [],
            "log": [],
        }
        _atomic_write(d / "session.json", data)
        return cls(sid)

    _summaries: dict = {}

    @staticmethod
    def _scan_all() -> list[tuple]:
        """(mtime, summary, slide times) of every tray, re-reading only the trays that changed."""
        out = []
        for f in sorted((library() / "sessions").glob("*/session.json"), reverse=True):
            try:
                mt = f.stat().st_mtime_ns
                hit = Session._summaries.get(str(f))
                if not hit or hit[0] != mt:
                    d = json.loads(f.read_text())
                    hit = (mt, summary(d), stats.slide_times(d))
                    Session._summaries[str(f)] = hit
            except Exception:
                continue
            out.append(hit)
        return out

    @staticmethod
    def list_all() -> list[dict]:
        return [hit[1] for hit in Session._scan_all()]

    @staticmethod
    def library_stats(target: int) -> dict:
        """Progress across every tray (stats.library_stats)."""
        all_ = Session._scan_all()
        return stats.library_stats([h[1] for h in all_], [t for h in all_ for t in h[2]], target)

    def save(self) -> None:
        with lock:
            _atomic_write(self.file, self.data)

    def group(self, gid: str) -> dict:
        for g in self.data["groups"]:
            if g["id"] == gid:
                return g
        raise KeyError(gid)

    def group_index(self, gid: str) -> int:
        return [g["id"] for g in self.data["groups"]].index(gid)

    def new_group(self, scans: list[str], rotation: int = 0, rot_reason: str = "") -> dict:
        return {
            "id": uuid.uuid4().hex[:8],
            "scans": scans,
            "excluded": [],
            "rotation": rotation,
            "rot_reason": rot_reason,
            "params": dict(self.data["defaults"]),
            "reviewed": False,
            "skip": False,
            "export": None,  # {"file","sha1","key"}
            "immich": None,  # {"asset_id","key"}
        }

    def log(self, msg: str) -> None:
        self.data.setdefault("log", []).append([time.time(), msg])
        self.data["log"] = self.data["log"][-200:]


def active_scans(g: dict) -> list[str]:
    s = [x for x in g["scans"] if x not in g.get("excluded", [])]
    return s or g["scans"][:1]


NEUTRAL_EXTRAS = {"curves": {}, "angle": 0.0, "crop": None}


def render_key(g: dict) -> str:
    """Identifies the exact output of a group; changes whenever the result would change."""
    # settings still at their neutral value are left out, so slides uploaded before a setting
    # existed (curves, straighten, crop) don't become "changed"
    params = {k: v for k, v in g["params"].items() if not (k in NEUTRAL_EXTRAS and v == NEUTRAL_EXTRAS[k])}
    k = json.dumps([active_scans(g), g["rotation"], params], sort_keys=True)
    return hashlib.sha1(k.encode()).hexdigest()[:12]


def tone_key(g: dict) -> str:
    """Identifies the tone curve's input (what its histogram shows): scans, restore, trim, geometry."""
    p = g["params"]
    k = json.dumps([active_scans(g), g["rotation"], p.get("strength"), p.get("trim"), p.get("angle", 0.0),
                    p.get("crop")])
    return hashlib.sha1(k.encode()).hexdigest()[:12]


# --------------------------------------------------------------------------- dates


DATE_RE = re.compile(r"^(\d{4})(?:-(\d{1,2})(?:-(\d{1,2}))?)?$")


def parse_date(v: str) -> tuple[datetime, int] | None:
    """'1978', '1978-06' or '1978-06-14' -> (start of that period, precision 1..3)."""
    m = DATE_RE.match((v or "").strip().replace("/", "-"))
    if not m:
        return None
    y, mo, d = int(m[1]), int(m[2] or 1), int(m[3] or 1)
    try:
        return datetime(y, max(1, min(12, mo)), max(1, min(31, d))), 1 + bool(m[2]) + bool(m[3])
    except ValueError:
        return None


def format_date(t: datetime, precision: int) -> str:
    return t.strftime(["%Y", "%Y-%m", "%Y-%m-%d"][precision - 1])


def slide_dates(d: dict) -> list[dict]:
    """The date each slide goes to Immich with, and where it came from.

    A slide's own date wins. Slides without one are estimated from the dated slides around them in
    tray order — a tray is one stretch of time, so slides between an August 1978 and a July 1979
    slide are interpolated between the two — then the tray's date, then (empty) the scan's EXIF."""
    groups = d["groups"]
    own = [parse_date(g.get("date", "")) for g in groups]
    dated = [i for i, x in enumerate(own) if x]
    tray = parse_date(d.get("date", ""))
    out = []
    for i, g in enumerate(groups):
        if own[i]:
            out.append({"value": format_date(*own[i]), "source": "own"})
            continue
        before = max((j for j in dated if j < i), default=None)
        after = min((j for j in dated if j > i), default=None)
        if before is not None and after is not None:
            (t0, p0), (t1, p1) = own[before], own[after]
            t = t0 + (t1 - t0) * ((i - before) / (after - before))
            out.append({"value": format_date(t, min(p0, p1)), "source": "between", "from": [before, after]})
        elif before is not None or after is not None:
            j = before if before is not None else after
            out.append({"value": format_date(*own[j]), "source": "near", "from": [j]})
        elif tray:
            out.append({"value": format_date(*tray), "source": "tray"})
        else:
            out.append({"value": "", "source": "scan"})
    return out


def meta_key(g: dict, date: dict) -> str:
    """What besides the pixels goes to Immich with a slide: its date and caption."""
    return hashlib.sha1(json.dumps([date.get("value", ""), g.get("caption", "")]).encode()).hexdigest()[:12]


def group_status(g: dict, meta: str | None = None) -> str:
    if g.get("skip"):
        return "skipped"
    im = g.get("immich")
    if im and g.get("locked"):
        return "uploaded"  # originals gone: Immich's copy is the final one
    if im and im.get("key") == render_key(g) and (meta is None or im.get("meta", meta) == meta):
        return "uploaded"
    if im:
        return "changed"
    return "reviewed" if g.get("reviewed") else "new"


def statuses(d: dict) -> list[str]:
    dates = slide_dates(d)
    return [group_status(g, meta_key(g, dt)) for g, dt in zip(d["groups"], dates)]


def summary(d: dict) -> dict:
    groups = d["groups"]
    st = statuses(d)
    return {
        "id": d["id"],
        "name": d["name"],
        "album": d["album"],
        "date": d.get("date", ""),
        "created": d["created"],
        "slides": len(groups),
        "scans": len(d["scans"]),
        "reviewed": sum(1 for g, s in zip(groups, st) if g.get("reviewed") or s in ("uploaded", "skipped")),
        "uploaded": st.count("uploaded"),
        "skipped": st.count("skipped"),
        "pending_upload": sum(1 for s in st if s in ("new", "reviewed", "changed")),
        # developed (marked ready) and not in Immich yet: what "upload the ready ones" sends
        "ready_upload": sum(1 for g, s in zip(groups, st) if g.get("reviewed") and s in ("reviewed", "changed")),
        "card_cleaned": d.get("card_cleaned", False),
        "sources": sorted({s.get("source_root", "") for s in d["scans"].values()}),
    }

File: frontend/src/standalone/store.ts
// Tray data and the keys derived from it (slidestation/store.py). session.json is the same file
// the Python app writes, and the keys are computed byte for byte like Python's, so a library
// folder can move between the browser and the desktop app without every slide turning "changed".
import type { GroupStatus, Params } from "@/lib/api";

export type Scan = {
  file: string;
  source: string;
  source_root: string;
  removable: boolean;
  size: number;
  sha1: string;
  taken: string;
  source_deleted: boolean;
  /** Pulled in from Immich: the asset it was downloaded from. */
  immich_asset?: string;
};

/** A slide's copy in Immich (see ARCHITECTURE "Round trip with Immich"). */
export type ImmichRecord = {
  asset_id: string;
  key: string;
  status?: string;
  meta?: string;
  /** What Immich was told besides the pixels: the day and the caption. */
  pushed?: { date: string; caption: string };
  /** The untouched scans stacked under it: scan -> asset, the stack, and which ones this app uploaded. */
  stack_id?: string;
  originals?: Record<string, string>;
  own_originals?: string[];
  /** When it went up (seconds), for the stats. */
  at?: number;
};

export type Snapshot = {
  params: Params;
  rotation: number;
  rot_reason: string;
  params_source: string;
  what?: string;
  t?: number;
};

export type GroupData = {
  id: string;
  scans: string[];
  excluded: string[];
  rotation: number;
  rot_reason: string;
  params: Params;
  reviewed: boolean;
  skip: boolean;
  export: { file: string; key: string; ekey: string; sha1: string } | null;
  immich: ImmichRecord | null;
  /** Pulled in from Immich: its upload replaces this asset. */
  source_asset?: { id: string };
  /** When it was marked developed (seconds), for the stats. */
  developed_at?: number;
  params_source?: string;
  auto_excluded?: Record<string, string>;
  locked?: string;
  date?: string;
  caption?: string;
  feat?: number[];
  history?: { undo: Snapshot[]; redo: Snapshot[] };
};

export type SessionData = {
  id: string;
  name: string;
  album: string;
  date: string;
  created: number;
  defaults: Params;
  scans: Record<string, Scan>;
  groups: GroupData[];
  log: [number, string][];
  card_cleaned?: boolean;
  immich_album_id?: string;
  date_key?: string;
  orphan_assets?: string[];
  orphan_stacks?: string[];
};

// ------------------------------------------------------------------ Python-compatible JSON

/** A number json.dumps prints as a float ("0.0", "1e-05"), as Python holds params. */
export function pyFloat(x: number): string {
  if (!Number.isFinite(x)) return x > 0 ? "Infinity" : x < 0 ? "-Infinity" : "NaN";
  if (Number.isInteger(x) && Math.abs(x) < 1e16) return x.toFixed(1);
  const a = Math.abs(x);
  if (a >= 1e-4 && a < 1e16) return String(x);
  const [m, e] = x.toExponential().split("e");
  const exp = Number(e);
  return `${m}e${exp < 0 ? "-" : "+"}${String(Math.abs(exp)).padStart(2, "0")}`;
}

/** An integer for pyDumps (everything else numeric is printed as a float). */
export class PyInt {
  constructor(public n: number) {}
}

const escape = (s: string) =>
  '"' +
  s.replace(/[\\"\u0000-\u001f\u007f-￿]/g, (c) => {
    const special: Record<string, string> = {
      "\\": "\\\\",
      '"': '\\"',
      "\n": "\\n",
      "\r": "\\r",
      "\t": "\\t",
      "\b": "\\b",
      "\f": "\\f",
    };
    return special[c] ?? "\\u" + c.charCodeAt(0).toString(16).padStart(4, "0");
  }) +
  '"';

/** json.dumps(v, sort_keys=sortKeys) with Python's separators and ensure_ascii. */
export function pyDumps(v: unknown, sortKeys = false): string {
  if (v === null || v === undefined) return "null";
  if (v instanceof PyInt) return String(Math.trunc(v.n));
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") return pyFloat(v);
  if (typeof v === "string") return escape(v);
  if (Array.isArray(v)) return "[" + v.map((x) => pyDumps(x, sortKeys)).join(", ") + "]";
  const keys = Object.keys(v as object);
  if (sortKeys) keys.sort();
  return (
    "{" + keys.map((k) => `${escape(k)}: ${pyDumps((v as Record<string, unknown>)[k], sortKeys)}`).join(", ") + "}"
  );
}

/**
 * session.json as the Python app writes it (indent=1). Numbers inside params are written as
 * floats, because Python keeps them as floats and hashes them that way in render_key.
 */
export function dumpSession(d: unknown): string {
  const walk = (v: unknown, floats: boolean, ind: string): string => {
    if (v === null || v === undefined) return "null";
    if (typeof v === "number") return floats ? pyFloat(v) : Number.isFinite(v) ? JSON.stringify(v) : "null";
    if (typeof v === "boolean" || typeof v === "string") return typeof v === "string" ? escape(v) : String(v);
    const inner = ind + " ";
    if (Array.isArray(v))
      return v.length ? "[\n" + v.map((x) => inner + walk(x, floats, inner)).join(",\n") + "\n" + ind + "]" : "[]";
    const entries = Object.entries(v as object).filter(([, x]) => x !== undefined);
    if (!entries.length) return "{}";
    return (
      "{\n" +
      entries
        .map(([k, x]) => `${inner}${escape(k)}: ${walk(x, floats || k === "params" || k === "defaults", inner)}`)
        .join(",\n") +
      "\n" +
      ind +
      "}"
    );
  };
  return walk(d, false, "");
}

// ------------------------------------------------------------------ SHA-1 (sync, for short keys)

export function sha1Hex(text: string): string {
  const bytes = new TextEncoder().encode(text);
  const len = bytes.length;
  const words = new Uint32Array((((len + 8) >> 6) + 1) * 16);
  for (let i = 0; i < len; i++) words[i >> 2] |= bytes[i] << (24 - (i % 4) * 8);
  words[len >> 2] |= 0x80 << (24 - (len % 4) * 8);
  words[words.length - 1] = len * 8;
  let [h0, h1, h2, h3, h4] = [0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0];
  const w = new Uint32Array(80);
  for (let off = 0; off < words.length; off += 16) {
    for (let t = 0; t < 16; t++) w[t] = words[off + t];
    for (let t = 16; t < 80; t++) {
      const x = w[t - 3] ^ w[t - 8] ^ w[t - 14] ^ w[t - 16];
      w[t] = (x << 1) | (x >>> 31);
    }
    let [a, b, c, d, e] = [h0, h1, h2, h3, h4];
    for (let t = 0; t < 80; t++) {
      const f = t < 20 ? (b & c) | (~b & d) : t < 40 ? b ^ c ^ d : t < 60 ? (b & c) | (b & d) | (c & d) : b ^ c ^ d;
      const k = t < 20 ? 0x5a827999 : t < 40 ? 0x6ed9eba1 : t < 60 ? 0x8f1bbcdc : 0xca62c1d6;
      const tmp = (((a << 5) | (a >>> 27)) + f + e + k + w[t]) >>> 0;
      e = d;
      d = c;
      c = (b << 30) | (b >>> 2);
      b = a;
      a = tmp;
    }
    h0 = (h0 + a) >>> 0;
    h1 = (h1 + b) >>> 0;
    h2 = (h2 + c) >>> 0;
    h3 = (h3 + d) >>> 0;
    h4 = (h4 + e) >>> 0;
  }
  return [h0, h1, h2, h3, h4].map((x) => (x >>> 0).toString(16).padStart(8, "0")).join("");
}

// ------------------------------------------------------------------ keys and statuses

export function activeScans(g: GroupData): string[] {
  const s = g.scans.filter((x) => !(g.excluded ?? []).includes(x));
  return s.length ? s : g.scans.slice(0, 1);
}

const isNeutral = (k: string, v: unknown) =>
  (k === "curves" && v && typeof v === "object" && !Object.keys(v).length) ||
  (k === "angle" && v === 0) ||
  (k === "crop" && v === null);

/** Identifies the exact output of a slide; changes whenever the result would change. */
export function renderKey(g: GroupData): string {
  // settings still at their neutral value are left out, so slides uploaded before a setting
  // existed (curves, straighten, crop) don't become "changed"
  const params = Object.fromEntries(Object.entries(g.params).filter(([k, v]) => !isNeutral(k, v)));
  return sha1Hex(pyDumps([activeScans(g), new PyInt(g.rotation), params], true)).slice(0, 12);
}

/** Identifies the tone curve's input (what its histogram shows). */
export function toneKey(g: GroupData): string {
  const p = g.params;
  return sha1Hex(
    pyDumps([activeScans(g), new PyInt(g.rotation), p.strength, p.trim, p.angle ?? 0, p.crop ?? null]),
  ).slice(0, 12);
}

const DATE_RE = /^(\d{4})(?:-(\d{1,2})(?:-(\d{1,2}))?)?$/;

/** '1978', '1978-06' or '1978-06-14' -> (start of that period in UTC ms, precision 1..3). */
export function parseDate(v: string | undefined): [number, number] | null {
  const m = DATE_RE.exec((v ?? "").trim().replace(/\//g, "-"));
  if (!m) return null;
  const y = +m[1];
  const mo = Math.max(1, Math.min(12, +(m[2] ?? 1)));
  const d = Math.max(1, Math.min(31, +(m[3] ?? 1)));
  const t = Date.UTC(y, mo - 1, d);
  const back = new Date(t);
  if (back.getUTCDate() !== d || back.getUTCMonth() !== mo - 1) return null; // e.g. 31 June
  const dt = new Date(t);
  dt.setUTCFullYear(y); // years below 100
  return [dt.getTime(), 1 + (m[2] ? 1 : 0) + (m[3] ? 1 : 0)];
}

export function formatDate(t: number, precision: number): string {
  const d = new Date(t);
  const parts = [
    String(d.getUTCFullYear()).padStart(4, "0"),
    String(d.getUTCMonth() + 1).padStart(2, "0"),
    String(d.getUTCDate()).padStart(2, "0"),
  ];
  return parts.slice(0, precision).join("-");
}

export type DateEst = { value: string; source: "own" | "between" | "near" | "tray" | "scan"; from?: number[] };

/** The date each slide goes to Immich with, and where it came from (store.slide_dates). */
export function slideDates(d: SessionData): DateEst[] {
  const own = d.groups.map((g) => parseDate(g.date));
  const dated = own.flatMap((x, i) => (x ? [i] : []));
  const tray = parseDate(d.date);
  return d.groups.map((_, i) => {
    const mine = own[i];
    if (mine) return { value: formatDate(...mine), source: "own" };
    const before = dated.filter((j) => j < i).pop();
    const after = dated.find((j) => j > i);
    if (before !== undefined && after !== undefined) {
      const [t0, p0] = own[before]!;
      const [t1, p1] = own[after]!;
      const t = t0 + (t1 - t0) * ((i - before) / (after - before));
      return { value: formatDate(t, Math.min(p0, p1)), source: "between", from: [before, after] };
    }
    const j = before ?? after;
    if (j !== undefined) return { value: formatDate(...own[j]!), source: "near", from: [j] };
    if (tray) return { value: formatDate(...tray), source: "tray" };
    return { value: "", source: "scan" };
  });
}

/** What besides the pixels goes to Immich with a slide: its date and caption. */
export const metaKey = (g: GroupData, date: DateEst) =>
  sha1Hex(pyDumps([date.value ?? "", g.caption ?? ""])).slice(0, 12);

export function groupStatus(g: GroupData, meta?: string): GroupStatus {
  if (g.skip) return "skipped";
  const im = g.immich;
  if (im && g.locked) return "uploaded"; // originals gone: Immich's copy is the final one
  if (im && im.key === renderKey(g) && (meta === undefined || (im.meta ?? meta) === meta)) return "uploaded";
  if (im) return "changed";
  return g.reviewed ? "reviewed" : "new";
}

export function statuses(d: SessionData): GroupStatus[] {
  const dates = slideDates(d);
  return d.groups.map((g, i) => groupStatus(g, metaKey(g, dates[i])));
}

export function summary(d: SessionData) {
  const st = statuses(d);
  const g = d.groups;
  return {
    id: d.id,
    name: d.name,
    album: d.album,
    date: d.date ?? "",
    created: d.created,
    slides: g.length,
    scans: Object.keys(d.scans).length,
    reviewed: g.filter((x, i) => x.reviewed || st[i] === "uploaded" || st[i] === "skipped").length,
    uploaded: st.filter((s) => s === "uploaded").length,
    skipped: st.filter((s) => s === "skipped").length,
    pending_upload: st.filter((s) => s === "new" || s === "reviewed" || s === "changed").length,
    // developed (marked ready) and not in Immich yet: what "upload the ready ones" sends
    ready_upload: g.filter((x, i) => x.reviewed && (st[i] === "reviewed" || st[i] === "changed")).length,
    card_cleaned: d.card_cleaned ?? false,
    sources: [...new Set(Object.values(d.scans).map((s) => s.source_root ?? ""))].sort(),
  };
}

export const slugify = (s: string) =>
  s
    .trim()
    .toLowerCase()
    .replace(/[^\p{L}\p{N}_-]+/gu, "-")
    .replace(/^-+|-+$/g, "") || "session";

export const randomHex = (n: number) =>
  Array.from(crypto.getRandomValues(new Uint8Array(Math.ceil(n / 2))), (b) => b.toString(16).padStart(2, "0"))
    .join("")
    .slice(0, n);

export function newGroup(d: SessionData, scans: string[], rotation = 0, rotReason = ""): GroupData {
  return {
    id: randomHex(8),
    scans,
    excluded: [],
    rotation,
    rot_reason: rotReason,
    params: structuredClone(d.defaults),
    reviewed: false,
    skip: false,
    export: null,
    immich: null,
  };
}
